File: Programmers/Exhaustive Search/피로도.py
from collections import deque
def solution(k, dungeons):
    answer = 0
    dq = deque()
    dq.append([k, []])

    while dq:
        fatigue, search = dq.popleft()
        for i in range(len(dungeons)):
            [a, b] = dungeons[i]
            if i not in search and fatigue >= a and fatigue - b >= 1:
                dq.append([fatigue - b, search + [i]])
            else:
                answer = max(answer, len(search))

        print(dq, answer)


    return answer

File: Programmers/피로도.java
public class 피로도 {
    int answer = -1;
    boolean[] visit;
    int[][] dungeons;

    public int solution(int k, int[][] dungeons) {
        this.dungeons = dungeons;
        visit = new boolean[dungeons.length];
        for(int i = 0; i < dungeons.length; i++) {
            if(k >= dungeons[i][0]) {
                dfs(i, k, 1);
            }
        }

        return answer;
    }

    private void dfs(int cur, int tired, int depth) {
        visit[cur] = true;
        tired -= dungeons[cur][1];
        for(int i = 0; i < dungeons.length; i++) {
            if(!visit[i] && dungeons[i][0] <= tired) {
                dfs(i, tired, depth + 1);
            }
        }
        answer = Math.max(answer, depth);
        visit[cur] = false;
    }
}
